File: packages/go/qorechain/unified/unified.go
// Package unified derives a single eth-native secp256k1 keypair that renders as
// the SAME 20-byte account under all THREE QoreChain address encodings, so a
// wallet never "has funds on the native lane but not the EVM lane" again.
//
// The 20 bytes are the Ethereum derivation keccak256(uncompressedPubKey[1:])[12:],
// so the key is natively spendable on the EVM lane; the native ("qor1…") and SVM
// (base58) forms are just other encodings of those same 20 bytes — the chain
// reads one x/bank balance for the account, visible under all three:
//
//	UnifiedAccount{
//	  PrivateKey   ([]byte, 32B),
//	  PublicKey    ([]byte, 33B compressed),
//	  AddressBytes ([20]byte),
//	  Cosmos       "qor1…"        (bech32),
//	  Evm          "0x…"          (EIP-55),
//	  Svm          "<base58>"     (base58(20B ‖ 12 zero bytes) = 32-byte SVM addr),
//	  Pqc          {PublicKey, SecretKey}  // ML-DSA-87 (Dilithium-5) for the hybrid ante
//	}
//
// Eth-native accounts are fully supported by the chain: its ante SigVerification
// handles eth_secp256k1 and the PQC hybrid decorator keys off the address, so the
// same wallet signs EVM txs AND PQC-hybrid native txs.
//
// This is ADDITIVE. The classic coin-118 native derivation in the accounts
// package (m/44'/118'/0'/0/i, ripemd160(sha256(pubkey))) is unchanged.
package unified

import (
	"encoding/hex"
	"errors"
	"fmt"

	"github.com/btcsuite/btcd/btcutil/bech32"
	"github.com/cloudflare/circl/sign/mldsa/mldsa87"
	"github.com/cosmos/go-bip39"
	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/mr-tron/base58"
	"golang.org/x/crypto/sha3"

	"github.com/qorechain/qorechain-sdk/packages/go/qorechain/accounts"
)

// NativePrefix is the bech32 human-readable prefix for QoreChain accounts.
const NativePrefix = "qor"

// coinTypeEVM is the Ethereum SLIP-0044 coin type; the unified account uses the
// eth HD path m/44'/60'/0'/0/{index} so the 20-byte address is EVM-native.
const coinTypeEVM uint32 = 60

// PQCKeypair is the ML-DSA-87 (Dilithium-5) half of a unified account. Treat
// SecretKey as a secret.
type PQCKeypair struct {
	PublicKey []byte // 2592 bytes
	SecretKey []byte // 4896 bytes
}

// UnifiedAccount is one eth-native secp256k1 identity rendered under all three
// QoreChain address encodings, plus its deterministic ML-DSA-87 hybrid key.
// Treat PrivateKey and Pqc.SecretKey as secrets.
type UnifiedAccount struct {
	PrivateKey   []byte     // 32-byte secp256k1 private key
	PublicKey    []byte     // 33-byte compressed secp256k1 public key
	AddressBytes [20]byte   // keccak256(uncompressedPubKey[1:])[12:]
	Cosmos       string     // bech32("qor", addr20)
	Evm          string     // EIP-55 mixed-case checksum "0x…"
	Svm          string     // base58(addr20 ‖ 12 zero bytes)
	Pqc          PQCKeypair // ML-DSA-87 hybrid key
}

// DeriveUnifiedAccount derives a unified eth-native account from a BIP-39
// mnemonic.
//
// HD path: m/44'/60'/0'/0/{index}, secp256k1. The 20-byte account address is
// keccak256(uncompressedPubKey[1:])[12:] (the Ethereum derivation). The PQC key
// is seeded from shake256("qorechain:pqc:v1|"+cosmos+"|"+mnemonic) so it is
// deterministic and recoverable from {address, mnemonic}.
func DeriveUnifiedAccount(mnemonic string, index uint32) (UnifiedAccount, error) {
	if !accounts.ValidateMnemonic(mnemonic) {
		return UnifiedAccount{}, errors.New("invalid mnemonic")
	}
	seed, err := bip39.NewSeedWithErrorChecking(mnemonic, "")
	if err != nil {
		return UnifiedAccount{}, err
	}
	privKey, compressed, addr20, err := deriveEth(seed, index)
	if err != nil {
		return UnifiedAccount{}, err
	}
	enc, err := AddressesFrom20(addr20)
	if err != nil {
		return UnifiedAccount{}, err
	}
	pqcSeed := shake256([]byte("qorechain:pqc:v1|"+enc.Cosmos+"|"+mnemonic), 32)
	pqc, err := pqcFromSeed(pqcSeed)
	if err != nil {
		return UnifiedAccount{}, err
	}
	return UnifiedAccount{
		PrivateKey:   privKey,
		PublicKey:    compressed,
		AddressBytes: addr20,
		Cosmos:       enc.Cosmos,
		Evm:          enc.Evm,
		Svm:          enc.Svm,
		Pqc:          pqc,
	}, nil
}

// UnifiedAccountFromSeed builds a unified account directly from a 32-byte seed
// used as the secp256k1 private key (no HD derivation).
//
// The PQC seed is shake256("qorechain:pqc:v1|"+cosmos+"|seed:"+hex(seed32)) —
// the literal "seed:" prefix plus the lowercase-hex 32-byte private key stands
// in for the mnemonic, matching the wallet adapter's walletFromSeed /
// fromPhantomSignature contract.
func UnifiedAccountFromSeed(seed32 []byte) (UnifiedAccount, error) {
	if len(seed32) != 32 {
		return UnifiedAccount{}, fmt.Errorf("seed must be 32 bytes, got %d", len(seed32))
	}
	priv := secp256k1.PrivKeyFromBytes(seed32)
	compressed := priv.PubKey().SerializeCompressed()
	uncompressed := priv.PubKey().SerializeUncompressed()
	var addr20 [20]byte
	copy(addr20[:], keccak(uncompressed[1:])[12:])

	enc, err := AddressesFrom20(addr20)
	if err != nil {
		return UnifiedAccount{}, err
	}
	pqcSeed := shake256([]byte("qorechain:pqc:v1|"+enc.Cosmos+"|seed:"+hex.EncodeToString(seed32)), 32)
	pqc, err := pqcFromSeed(pqcSeed)
	if err != nil {
		return UnifiedAccount{}, err
	}
	privCopy := make([]byte, 32)
	copy(privCopy, seed32)
	return UnifiedAccount{
		PrivateKey:   privCopy,
		PublicKey:    compressed,
		AddressBytes: addr20,
		Cosmos:       enc.Cosmos,
		Evm:          enc.Evm,
		Svm:          enc.Svm,
		Pqc:          pqc,
	}, nil
}

// Addresses holds the three address encodings of a single 20-byte account.
type Addresses struct {
	AddressBytes [20]byte
	Cosmos       string // bech32("qor", addr20)
	Evm          string // EIP-55 "0x…"
	Svm          string // base58(addr20 ‖ 12 zero bytes)
}

// AddressesFrom20 renders the three encodings of a known 20-byte account
// address. Exposed so SDKs/backends can render all three from an account too.
func AddressesFrom20(addr20 [20]byte) (Addresses, error) {
	words, err := bech32.ConvertBits(addr20[:], 8, 5, true)
	if err != nil {
		return Addresses{}, fmt.Errorf("convert bech32 words: %w", err)
	}
	cosmos, err := bech32.Encode(NativePrefix, words)
	if err != nil {
		return Addresses{}, fmt.Errorf("encode bech32: %w", err)
	}
	svmBytes := make([]byte, 32)
	copy(svmBytes, addr20[:]) // right-pad with 12 zero bytes → unified 32-byte SVM address
	return Addresses{
		AddressBytes: addr20,
		Cosmos:       cosmos,
		Evm:          toEIP55(addr20[:]),
		Svm:          base58.Encode(svmBytes),
	}, nil
}

// QoreAddresses returns the three encodings of an existing account, given
// exactly one of a bech32 cosmos address, an EVM hex address, or a raw hex
// address (all forms of the same 20 bytes).
func QoreAddresses(cosmos, evm, hexAddr string) (Addresses, error) {
	var addr20 [20]byte
	switch {
	case evm != "":
		b, err := decodeHex20(evm)
		if err != nil {
			return Addresses{}, err
		}
		addr20 = b
	case hexAddr != "":
		b, err := decodeHex20(hexAddr)
		if err != nil {
			return Addresses{}, err
		}
		addr20 = b
	case cosmos != "":
		_, words, err := bech32.Decode(cosmos)
		if err != nil {
			return Addresses{}, fmt.Errorf("invalid bech32 address: %w", err)
		}
		data, err := bech32.ConvertBits(words, 5, 8, false)
		if err != nil {
			return Addresses{}, fmt.Errorf("invalid bech32 payload: %w", err)
		}
		if len(data) != 20 {
			return Addresses{}, fmt.Errorf("bech32 payload must be 20 bytes, got %d", len(data))
		}
		copy(addr20[:], data)
	default:
		return Addresses{}, errors.New("provide one of {cosmos, evm, hex}")
	}
	return AddressesFrom20(addr20)
}

// --- internal helpers ---

// deriveEth derives the eth-path secp256k1 key and its 20-byte address from a
// BIP-39 seed. Path: m/44'/60'/0'/0/{index}.
func deriveEth(seed []byte, index uint32) (privKey, compressed []byte, addr20 [20]byte, err error) {
	master, err := newSecp256k1Master(seed)
	if err != nil {
		return nil, nil, addr20, err
	}
	node, err := master.derivePath([]uint32{
		44 + hardenedOffset,
		coinTypeEVM + hardenedOffset,
		0 + hardenedOffset,
		0,
		index,
	})
	if err != nil {
		return nil, nil, addr20, err
	}
	priv := secp256k1.PrivKeyFromBytes(node.key)
	compressed = priv.PubKey().SerializeCompressed()
	uncompressed := priv.PubKey().SerializeUncompressed()
	copy(addr20[:], keccak(uncompressed[1:])[12:])
	privKey = make([]byte, 32)
	copy(privKey, node.key)
	return privKey, compressed, addr20, nil
}

// pqcFromSeed runs ML-DSA-87 (Dilithium-5) seeded keygen from a 32-byte seed.
func pqcFromSeed(seed []byte) (PQCKeypair, error) {
	if len(seed) != 32 {
		return PQCKeypair{}, fmt.Errorf("pqc seed must be 32 bytes, got %d", len(seed))
	}
	var s [mldsa87.SeedSize]byte
	copy(s[:], seed)
	pub, priv := mldsa87.NewKeyFromSeed(&s)
	pubBytes, err := pub.MarshalBinary()
	if err != nil {
		return PQCKeypair{}, err
	}
	privBytes, err := priv.MarshalBinary()
	if err != nil {
		return PQCKeypair{}, err
	}
	return PQCKeypair{PublicKey: pubBytes, SecretKey: privBytes}, nil
}

func decodeHex20(s string) ([20]byte, error) {
	var out [20]byte
	body := s
	if len(body) >= 2 && (body[:2] == "0x" || body[:2] == "0X") {
		body = body[2:]
	}
	b, err := hex.DecodeString(body)
	if err != nil {
		return out, fmt.Errorf("invalid hex address: %w", err)
	}
	if len(b) != 20 {
		return out, fmt.Errorf("address must be 20 bytes, got %d", len(b))
	}
	copy(out[:], b)
	return out, nil
}

func keccak(data []byte) []byte {
	h := sha3.NewLegacyKeccak256()
	h.Write(data)
	return h.Sum(nil)
}

func shake256(data []byte, n int) []byte {
	out := make([]byte, n)
	sha3.ShakeSum256(out, data)
	return out
}

// toEIP55 renders 20 address bytes as a 0x-prefixed EIP-55 mixed-case checksum.
func toEIP55(addr []byte) string {
	lower := hex.EncodeToString(addr)
	hash := keccak([]byte(lower))
	out := make([]byte, 0, 42)
	out = append(out, '0', 'x')
	for i := 0; i < len(lower); i++ {
		c := lower[i]
		if c >= '0' && c <= '9' {
			out = append(out, c)
			continue
		}
		nibble := hash[i/2]
		if i%2 == 0 {
			nibble >>= 4
		} else {
			nibble &= 0x0f
		}
		if nibble >= 8 {
			out = append(out, c-'a'+'A')
		} else {
			out = append(out, c)
		}
	}
	return string(out)
}

File: packages/py/src/qorsdk/unified.py
"""Unified eth-native QoreChain wallet.

One eth-native secp256k1 keypair yields the SAME 20-byte account rendered as all
THREE QoreChain address encodings, so a wallet never "has funds on one lane but
not another" again. The 20 bytes are the Ethereum derivation
``keccak256(uncompressed_pubkey[1:])[12:]`` so the key is natively spendable on
the EVM lane; the native (``qor1…``) and SVM (base58) forms are just other
encodings of those same 20 bytes — the chain reads one x/bank balance for the
account, visible under all three.

    UnifiedAccount:
      private_key   (32 bytes),
      public_key    (33 bytes, compressed secp256k1),
      address_bytes (20 bytes),
      cosmos        "qor1…"            (bech32),
      evm           "0x…" (EIP-55),    (hex),
      svm           "<base58>"         (base58(20B ‖ 12 zero bytes) = 32-byte SVM addr),
      pqc           PqcKeypair         (ML-DSA-87, seeded + recoverable)

An eth-native account signs EVM txs (EIP-155) AND, via the Native-lane
``eth_secp256k1`` scheme (see :mod:`qorsdk.sign_eth`), PQC-hybrid Native txs —
one identity, all lanes.

This mirrors ``@qorechain/wallet-adapter`` (``wallet.js``) byte-for-byte but is
reimplemented natively here, with no dependency on the adapter. It is ADDITIVE:
the classical coin-118 native derivation in :mod:`qorsdk.accounts` is unchanged.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass

import base58
from bip_utils import Bip32Slip10Secp256k1, Bip39SeedGenerator
from ecdsa import SECP256k1, SigningKey

from .accounts import validate_mnemonic
from .address import DEFAULT_PREFIX
from .pqc import PqcKeypair
from .utils.hash import keccak256

try:  # dilithium-py exposes seeded keygen as ML_DSA_87.key_derive(seed32)
    from dilithium_py.ml_dsa import ML_DSA_87
except Exception as exc:  # pragma: no cover - dependency guaranteed by pyproject
    raise ImportError("dilithium-py is required for unified PQC derivation") from exc

#: Ethereum HD path (coin-type 60) — makes the 20-byte address the keccak
#: derivation, so the key is EVM-native (spendable via eth_sendRawTransaction).
_ETH_HD_PATH = "44'/60'/0'/0/{index}"

#: bech32 human-readable prefix for QoreChain account addresses.
HRP = DEFAULT_PREFIX


@dataclass(frozen=True)
class UnifiedAccount:
    """A single eth-native account rendered in every QoreChain encoding.

    ``private_key`` and ``pqc.secret_key`` are secret material — never log them.
    ``public_key`` is the 33-byte compressed secp256k1 key; ``address_bytes`` is
    the 20-byte Ethereum-derivation account address the three encodings share.
    """

    private_key: bytes
    public_key: bytes
    address_bytes: bytes
    cosmos: str
    evm: str
    svm: str
    pqc: PqcKeypair

    @property
    def private_key_hex(self) -> str:
        """The private key as a ``0x``-prefixed hex string."""
        return "0x" + self.private_key.hex()

    @property
    def public_key_hex(self) -> str:
        """The compressed public key as a ``0x``-prefixed hex string."""
        return "0x" + self.public_key.hex()


def _eip55(hex20: str) -> str:
    """EIP-55 mixed-case checksum for a 20-byte hex address (no ``0x``)."""
    lower = hex20.lower()
    digest = keccak256(lower.encode("ascii")).hex()
    out = "0x"
    for i, char in enumerate(lower):
        out += char.upper() if int(digest[i], 16) >= 8 else char
    return out


def _bech32(addr20: bytes, prefix: str = HRP) -> str:
    from bech32 import bech32_encode, convertbits

    words = convertbits(addr20, 8, 5, pad=True)
    if words is None:  # pragma: no cover - 20 bytes always converts
        raise ValueError("failed to convert address to bech32 words")
    encoded = bech32_encode(prefix, words)
    if encoded is None:  # pragma: no cover
        raise ValueError("failed to encode bech32 address")
    return encoded


def addresses_from_20(addr20: bytes) -> dict[str, str]:
    """Render the three QoreChain encodings from a 20-byte account address.

    Exposed so SDKs / backends can render all three from a known account too.

    :returns: ``{"cosmos", "evm", "svm"}`` — the bech32 (``qor``), EIP-55 hex,
        and base58 (32-byte, right-padded with 12 zero bytes) encodings.
    :raises ValueError: If ``addr20`` is not exactly 20 bytes.
    """
    if len(addr20) != 20:
        raise ValueError("address must be 20 bytes")
    svm_bytes = bytes(addr20) + bytes(12)  # right-pad → unified 32-byte SVM address
    return {
        "cosmos": _bech32(addr20),
        "evm": _eip55(addr20.hex()),
        "svm": base58.b58encode(svm_bytes).decode("ascii"),
    }


def _derive_pqc(cosmos_addr: str, secret_material: str) -> PqcKeypair:
    """Deterministic ML-DSA-87 keypair from ``{cosmos_addr, secret_material}``.

    ``secret_material`` is the full trailing segment of the domain-separated PQC
    seed string: the raw mnemonic (for mnemonic-derived accounts) or
    ``"seed:" + hex(privkey)`` (for seed-derived accounts). Seed =
    ``shake256("qorechain:pqc:v1|" + cosmos + "|" + secret_material, 32)``.
    """
    seed = hashlib.shake_256(
        b"qorechain:pqc:v1|" + cosmos_addr.encode("utf-8") + b"|" + secret_material.encode("utf-8")
    ).digest(32)
    public_key, secret_key = ML_DSA_87.key_derive(seed)
    return PqcKeypair(public_key=bytes(public_key), secret_key=bytes(secret_key))


def _account_from_privkey(private_key: bytes, secret_material: str) -> UnifiedAccount:
    """Build a :class:`UnifiedAccount` from a 32-byte secp256k1 private key."""
    vk = SigningKey.from_string(private_key, curve=SECP256k1).get_verifying_key()
    uncompressed = vk.to_string("raw")  # 64 bytes: X || Y (no 0x04 prefix)
    compressed = vk.to_string("compressed")  # 33 bytes
    addr20 = keccak256(uncompressed)[12:]  # keccak of the 64-byte X||Y
    enc = addresses_from_20(addr20)
    return UnifiedAccount(
        private_key=bytes(private_key),
        public_key=bytes(compressed),
        address_bytes=bytes(addr20),
        cosmos=enc["cosmos"],
        evm=enc["evm"],
        svm=enc["svm"],
        pqc=_derive_pqc(enc["cosmos"], secret_material),
    )


def derive_unified_account(mnemonic: str, index: int = 0) -> UnifiedAccount:
    """Derive a unified eth-native QoreChain account from a BIP-39 mnemonic.

    HD path ``m/44'/60'/0'/0/{index}`` (secp256k1). The 20-byte account address
    is the Ethereum derivation ``keccak256(uncompressed_pubkey[1:])[12:]``,
    rendered as the ``qor1…`` (bech32), ``0x…`` (EIP-55), and base58 (SVM)
    encodings. The PQC key is a deterministic ML-DSA-87 keypair recoverable from
    ``{cosmos_address, mnemonic}``.

    :param mnemonic: A valid BIP-39 mnemonic (validated; a typo'd phrase raises).
    :param index: The account index in the HD path (default ``0``).
    :raises ValueError: On an invalid mnemonic or a negative index.
    """
    if not validate_mnemonic(mnemonic):
        raise ValueError("invalid mnemonic")
    if not isinstance(index, int) or isinstance(index, bool) or index < 0:
        raise ValueError(f"index must be a non-negative integer, got {index}")
    seed = bytes(Bip39SeedGenerator(mnemonic).Generate())
    node = Bip32Slip10Secp256k1.FromSeed(seed).DerivePath(_ETH_HD_PATH.format(index=index))
    private_key = bytes(node.PrivateKey().Raw().ToBytes())
    return _account_from_privkey(private_key, mnemonic)


def unified_account_from_seed(seed32: bytes) -> UnifiedAccount:
    """Build a unified account from a 32-byte seed used directly as the privkey.

    ``seed32`` is used verbatim as the secp256k1 private key (no HD derivation).
    The PQC seed's trailing segment is ``"seed:" + seed32.hex()`` (lowercase, no
    ``0x``) in place of the mnemonic, so the account is fully recoverable from the
    32-byte seed alone.

    .. warning::

       ``seed32`` MUST be real secret entropy — a CSPRNG draw
       (``secrets.token_bytes(32)``) or key material the user already keeps
       secret. It becomes the account's spend key verbatim, so anyone who learns
       it controls the account. It must NEVER be derived from a wallet
       signature, nor from any other value a third party can ask the user's
       wallet (or the user) to produce: such a value is a bearer secret handed
       out on request, so the account it derives is only as private as the party
       that asked for it. To spend from an existing external key, link it with
       ``MsgRegisterAuthenticator`` and use the authenticator lanes
       (``MsgExecuteCosmos`` / ``MsgExecuteEVM``) instead of deriving a new key.

    :raises ValueError: If ``seed32`` is not exactly 32 bytes.
    """
    if len(seed32) != 32:
        raise ValueError("seed must be 32 bytes")
    return _account_from_privkey(bytes(seed32), "seed:" + seed32.hex())


#: The error raised by the removed signature-derivation entry point.
_SIGNATURE_DERIVATION_REMOVED = (
    "unified_account_from_phantom_signature was removed in v0.8.0: deriving a "
    "spend key from a wallet signature is unsafe — the signature is a bearer "
    "secret that any page can request from the wallet, so whoever obtains it "
    "controls the account. Use the authenticator lanes instead: register the "
    "external key with MsgRegisterAuthenticator and spend via MsgExecuteCosmos "
    "/ MsgExecuteEVM. Any account previously derived this way must be treated "
    "as exposed — move its funds."
)


def unified_account_from_phantom_signature(sig_bytes: bytes) -> UnifiedAccount:
    """REMOVED in v0.8.0 — always raises :class:`NotImplementedError`.

    This used to derive a unified account's spend key from a wallet signature.
    That is unsafe: the signature is a bearer secret the wallet will hand to any
    page that asks for it, so anyone who obtains it controls the derived
    account. The symbol is retained only so existing callers fail loudly instead
    of silently keeping an exposed account.

    Use the authenticator lanes instead — register the external key with
    ``MsgRegisterAuthenticator`` (see :mod:`qorsdk.authenticator`) and spend from
    the canonical account via ``MsgExecuteCosmos`` / ``MsgExecuteEVM``. Any
    account previously derived this way must be treated as exposed: move its
    funds.

    :raises NotImplementedError: Always.
    """
    raise NotImplementedError(_SIGNATURE_DERIVATION_REMOVED)


def qore_addresses(
    *, cosmos: str | None = None, evm: str | None = None, hex: str | None = None
) -> dict[str, str]:
    """Render the three encodings of an existing account from any one form.

    Provide exactly one of ``cosmos`` (a ``qor1…`` bech32), ``evm`` (a ``0x`` hex
    address), or ``hex`` (a bare/``0x`` 20-byte hex). Returns
    ``{"cosmos", "evm", "svm"}``.

    :raises ValueError: If none (or more than the payload of one) is provided, or
        the input does not decode to 20 bytes.
    """
    if evm is not None:
        addr20 = bytes.fromhex(evm[2:] if evm[:2].lower() == "0x" else evm)
    elif hex is not None:
        addr20 = bytes.fromhex(hex[2:] if hex[:2].lower() == "0x" else hex)
    elif cosmos is not None:
        from bech32 import bech32_decode, convertbits

        _hrp, words = bech32_decode(cosmos)
        if words is None:
            raise ValueError(f"invalid bech32 address: {cosmos}")
        data = convertbits(words, 5, 8, pad=False)
        if data is None:
            raise ValueError(f"invalid bech32 payload: {cosmos}")
        addr20 = bytes(data)
    else:
        raise ValueError("provide one of {cosmos, evm, hex}")
    return addresses_from_20(addr20)
